File: autocontext/src/autocontext/execution/queued_task_browser_context.py
"""Browser-backed reference-context enrichment for queued tasks."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from autocontext.config.settings import AppSettings
from autocontext.integrations.browser.context_capture import (
    capture_browser_context,
    render_captured_browser_context,
)


class QueuedTaskBrowserContextService(Protocol):
    """Build authoritative reference context from a browser snapshot."""

    def build_reference_context(
        self,
        *,
        task_id: str,
        browser_url: str,
        reference_context: str | None,
    ) -> str: ...


@dataclass(frozen=True, slots=True)
class SettingsBackedQueuedTaskBrowserContextService:
    """Capture browser context for queued tasks using AppSettings."""

    settings: AppSettings

    def build_reference_context(
        self,
        *,
        task_id: str,
        browser_url: str,
        reference_context: str | None,
    ) -> str:
        context = capture_browser_context(
            self.settings,
            browser_url=browser_url,
            evidence_root=(self.settings.runs_root / "task_queue" / task_id),
        )
        return merge_queued_task_reference_context(
            reference_context=reference_context,
            browser_context=render_captured_browser_context(context),
        )


def create_queued_task_browser_context_service(
    settings: AppSettings,
) -> QueuedTaskBrowserContextService:
    """Create a settings-backed queued-task browser context service."""
    return SettingsBackedQueuedTaskBrowserContextService(settings=settings)


def merge_queued_task_reference_context(
    *,
    reference_context: str | None,
    browser_context: str,
) -> str:
    """Merge queued-task reference context with browser-derived context."""
    parts = []
    trimmed_reference_context = (reference_context or "").strip()
    if trimmed_reference_context:
        parts.append(trimmed_reference_context)
    trimmed_browser_context = browser_context.strip()
    if trimmed_browser_context:
        parts.append(trimmed_browser_context)
    return "\n\n".join(parts)


__all__ = [
    "QueuedTaskBrowserContextService",
    "SettingsBackedQueuedTaskBrowserContextService",
    "create_queued_task_browser_context_service",
    "merge_queued_task_reference_context",
]

File: ts/src/execution/queued-task-browser-context.ts
import { join, resolve } from "node:path";

import {
  captureBrowserContextFromUrl,
  renderCapturedBrowserContext,
  type BrowserContextCaptureSettingsLike,
} from "../integrations/browser/context-capture.js";

export interface QueuedTaskBrowserContextRequest {
  readonly taskId: string;
  readonly browserUrl: string;
  readonly referenceContext?: string;
}

export interface QueuedTaskBrowserContextService {
  buildReferenceContext(request: QueuedTaskBrowserContextRequest): Promise<string>;
}

export interface QueuedTaskBrowserContextSettingsLike extends BrowserContextCaptureSettingsLike {
  readonly runsRoot: string;
}

export interface QueuedTaskBrowserContextDependencies {
  readonly captureBrowserContextFromUrl: typeof captureBrowserContextFromUrl;
}

const DEFAULT_DEPENDENCIES: QueuedTaskBrowserContextDependencies = {
  captureBrowserContextFromUrl,
};

export class SettingsBackedQueuedTaskBrowserContextService implements QueuedTaskBrowserContextService {
  readonly #settings: QueuedTaskBrowserContextSettingsLike;
  readonly #dependencies: QueuedTaskBrowserContextDependencies;

  constructor(
    settings: QueuedTaskBrowserContextSettingsLike,
    dependencies: QueuedTaskBrowserContextDependencies = DEFAULT_DEPENDENCIES,
  ) {
    this.#settings = settings;
    this.#dependencies = dependencies;
  }

  async buildReferenceContext(request: QueuedTaskBrowserContextRequest): Promise<string> {
    const context = await this.#dependencies.captureBrowserContextFromUrl({
      settings: this.#settings,
      browserUrl: request.browserUrl,
      evidenceRoot: join(resolve(this.#settings.runsRoot), "task_queue", request.taskId),
    });
    return mergeQueuedTaskReferenceContext(
      request.referenceContext,
      renderCapturedBrowserContext(context),
    );
  }
}

export function createQueuedTaskBrowserContextService(
  settings: QueuedTaskBrowserContextSettingsLike,
  dependencies: QueuedTaskBrowserContextDependencies = DEFAULT_DEPENDENCIES,
): QueuedTaskBrowserContextService {
  return new SettingsBackedQueuedTaskBrowserContextService(settings, dependencies);
}

export function mergeQueuedTaskReferenceContext(
  referenceContext: string | undefined,
  browserContext: string,
): string {
  const trimmedReferenceContext = referenceContext?.trim();
  const trimmedBrowserContext = browserContext.trim();
  return [trimmedReferenceContext, trimmedBrowserContext].filter(Boolean).join("\n\n");
}
